### ISSUE TITLE
when pnpm-lock.yaml provides invalid resolutions, the store is corrupted resulting in ENOENT "file not found" errors

### ISSUE THREAD

**Issue by aquark** [pnpm/pnpm#2076]
### Code to reproduce the issue:

We have a package whose `pnpm-lock.yaml` was somehow corrupted so that certain packages were resolved to the wrong tarballs. Still not sure how it got corrupted, but the corruption wasn't detected as I'd expect. Example (private repo details elided):

```
  /tcomb/3.2.29:
    dev: false
    resolution:
      integrity: sha512-w2VwSrBoHa5BsSyH+KxEqeQBAllHhccyMFVHtGtdMpF4W7IRWfZjFiQceJPChOeTsSDVUpER2T8FA93pr0L+QA==
      tarball: '.../tar/-/tar-4.4.13.tgz'
```

As a result, when we `pnpm install` this package, the wrong files end up in the store, e.g. `.../.pnpm-store/2/.../tcomb/3.2.29/` contains the artifacts of `tar-4.4.13.tgz`, e.g. `node_modules/tar`, and this in turn results in errors like this:

`ERROR  ENOENT: no such file or directory, link '.../.pnpm-store/2/.../tcomb/3.2.29/node_modules/tcomb/LICENSE' -> '.../node_modules/.../tcomb/3.2.29/node_modules/_tmp_87507_ed43b3c452a6f106e35ab6f7bba5bcbf/LICENSE'`

### Expected behavior:

When resolving a package from a lockfile, there should be a sanity check that the resolved target actually points to the right package name and version, and if not, the lockfile should be rejected.

### Actual behavior:

Confusing `ENOENT` error which doesn't indicate the source of the problem.

### Additional information:

 - `node -v` prints: `v12.4.0`
 - Windows, OS X, or Linux?: macOS and Ubuntu Linux

**Comment by aquark** [comment]
Looking at the packages involved, I'm pretty sure that the corruption was the result of a bad merge conflict resolution.

**Comment by zkochan** [comment]
We'll add a custom merge resolver to pnpm: #2036

**Comment by zkochan** [comment]
I think there is much less chance this will happen with the additional verifications published in [v5.18.5](https://github.com/pnpm/pnpm/releases/tag/v5.18.5)

**Comment by zkochan** [comment]
I believe we have fixed this in some of the latest v6 versions.
